crates/rust-analyzer/src/main_loop.rs

```diff
@@ -1019,6 +1022,19 @@ impl GlobalState {
                     Some(format!("rust-analyzer/flycheck/{id}")),
                 );
             }
+            FlycheckMessage::ClearCrateDiagnostics { id, generation, package_id } => {
+                let snap = self.snapshot();
+                let Ok(Some(crate_id)) = snap.analysis.crate_with_id(&package_id) else {
```

Comment:
Here we should instead iterate the workspaces (and then the packages in those) to find the relevant crates